motion/thruster_interface_auv_cpp/src/thruster_interface_auv_driver.cpp

```diff
@@ -0,0 +1,138 @@
+#include "thruster_interface_auv_driver.hpp"
+
+ThrusterInterfaceAUVDriver::ThrusterInterfaceAUVDriver(){}
+
+ThrusterInterfaceAUVDriver::ThrusterInterfaceAUVDriver(
+    int I2C_BUS,
+    int PICO_I2C_ADDRESS,
+    double SYSTEM_OPERATIONAL_VOLTAGE,
+    const std::vector<int>& THRUSTER_MAPPING,
+    const std::vector<int>& THRUSTER_DIRECTION,
+    const std::vector<int>& THRUSTER_PWM_OFFSET,
+    const std::vector<int>& PWM_MIN,
+    const std::vector<int>& PWM_MAX,
+    const std::map<int, std::map<std::string, std::vector<double>>>& COEFFS
+) : I2C_BUS(I2C_BUS),
+    PICO_I2C_ADDRESS(PICO_I2C_ADDRESS),
+    //SYSTEM_OPERATIONAL_VOLTAGE(SYSTEM_OPERATIONAL_VOLTAGE), done after in the if-else
+    THRUSTER_MAPPING(THRUSTER_MAPPING),
+    THRUSTER_DIRECTION(THRUSTER_DIRECTION),
+    THRUSTER_PWM_OFFSET(THRUSTER_PWM_OFFSET),
+    PWM_MIN(PWM_MIN),
+    PWM_MAX(PWM_MAX),
+    COEFFS(COEFFS)
+{
+
+    // Convert SYSTEM_OPERATIONAL_VOLTAGE passed as argument to assign the integer field variable [one is double, field var is int]
+    if (SYSTEM_OPERATIONAL_VOLTAGE < 11.0) {
+        this->SYSTEM_OPERATIONAL_VOLTAGE = 10;
+    } else if (SYSTEM_OPERATIONAL_VOLTAGE < 13.0) {
+        this->SYSTEM_OPERATIONAL_VOLTAGE = 12;
+    } else if (SYSTEM_OPERATIONAL_VOLTAGE < 15.0) {
+        this->SYSTEM_OPERATIONAL_VOLTAGE = 14;
+    } else if (SYSTEM_OPERATIONAL_VOLTAGE < 17.0) {
+        this->SYSTEM_OPERATIONAL_VOLTAGE = 16;
+    } else if (SYSTEM_OPERATIONAL_VOLTAGE < 19.0) {
+        this->SYSTEM_OPERATIONAL_VOLTAGE = 18;
+    } else {
+        this->SYSTEM_OPERATIONAL_VOLTAGE = 20;
+    }
+
+    // Open the I2C bus
+    std::string i2c_filename = "/dev/i2c-" + std::to_string(I2C_BUS);
+    bus_fd = open(i2c_filename.c_str(), O_RDWR);
+    if (bus_fd < 0) {
+        std::cerr << "ERROR: Failed to open I2C bus " << I2C_BUS << std::endl;
+    }
+}
+
+ThrusterInterfaceAUVDriver::~ThrusterInterfaceAUVDriver() {
+    if (bus_fd >= 0) {
+        close(bus_fd);
+    }
+}
+
+std::vector<int16_t> ThrusterInterfaceAUVDriver::interpolate_forces_to_pwm(const std::vector<double>& thruster_forces_array) {
+    // Convert Newtons to Kg (since the thruster datasheet is in Kg)
+    std::vector<double> forces_in_kg(thruster_forces_array.size());
+    for (size_t i = 0; i < thruster_forces_array.size(); ++i) {
+        forces_in_kg[i] = thruster_forces_array[i] / 9.80665;
+    }
+
+    // Select the appropriate coefficients based on the operational voltage
+    auto left_coeffs = COEFFS[SYSTEM_OPERATIONAL_VOLTAGE]["LEFT"];
```

Comment:
Try to avoid using auto in general. Especially when it is not easy to deduce what type it is

Reply:
I suggest following [ES.11: Use auto to avoid redundant repetition of type names](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es11-use-auto-to-avoid-redundant-repetition-of-type-names)

Reply:
that variable should be a std::vector<double>, like the other parameters passed as std::vector<int>. The problem is that .as_integer_array() methods returns a long int and to do what we want to do i think it would require an explicit cast from long int to int that, imo, looks like inelegant and verbose. So i use auto. If i try to compile using explicit typing i get this:

> error: conversion from ‘vector<long int,allocator<long int>>’ to non-scalar type ‘vector<int,allocator<int>>’ requested
>    34 |         this->get_parameter("propulsion.thrusters.thruster_to_pin_mapping")
>       |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>    35 |             .as_integer_array();

Reply:
To be fair, if you end up not caring about system operational voltage, you can probably instantiate the left/right coefficients in the constructor?

Also, if you implement my suggested change and let your input be a parameter struct of some sorts, you can do this casting / validation immediately in the constructor.

Note that while long int -> int is a narrowing conversion, the thruster-to-pin mapping int will never be larger than what a normal int can hold, so I would actually suggest doing this cast

Reply:
conversion is done when passing them to the driver class
```
// Initialize thruster driver
    this->thruster_driver_ = ThrusterInterfaceAUVDriver(
        i2c_bus, i2c_address,
        std::vector<short>(thruster_mapping.begin(), thruster_mapping.end()),
        std::vector<short>(thruster_direction.begin(),\
                           thruster_direction.end()),\
        std::vector<int>(thruster_PWM_min.begin(), thruster_PWM_min.end()),
        std::vector<int>(thruster_PWM_max.begin(), thruster_PWM_max.end()),
        std::vector<double>(left_coeffs.begin(), left_coeffs.end()),
        std::vector<double>(right_coeffs.begin(), right_coeffs.end()));
```